### ISSUE TITLE
Support for Inline Calculations in `aes()`

### ISSUE THREAD

**Issue by atantos** [TidierOrg/TidierPlots.jl#49]
Hi.

 I'd like to suggest a feature enhancement to `TidierPlots.jl` that would improve its usability, especially for users transitioning from R's ggplot2.

In ggplot2, it is possible to perform inline calculations within the `aes()` function, such as `ggplot(aes(x = candidatevotes / totalvotes, fill = party))`. This feature allows for dynamic and on-the-fly transformations of data, making the plotting process more flexible and powerful.

As of now, `TidierPlots.jl` requires pre-calculation of such transformations before passing them to `aes()`. This approach, while functional, lacks the convenience and elegance of inline calculations.

This would align `TidierPlots.jl` more closely with ggplot2's capabilities and provide a more seamless experience for users familiar with ggplot2 syntax.

Thanks!

**Comment by rdboyes** [comment]
We should be able to support this using some of the parsing capability @kdpsingh developed for TidierData! Stay tuned

**Comment by rdboyes** [comment]
@kdpsingh this would require using TidierData.parse_tidy I think, which isn't currently exported. I remember there was talk of moving the parsing functionality to its own sub-package - is that still the plan?

**Comment by kdpsingh** [comment]
Hi @rdboyes and @atantos, thanks for tagging me on this.

While I was originally thinking of removing `parse_tidy()` from the TidierData package, the reason I haven't done it yet is that `parse_tidy()` converts tidy expressions into expressions that can be directly interpolated into DataFrames functions, so the functionality is specific to DataFrames (and not Julia more broadly). While AlgebraOfGraphics.jl has its own mini-language for data transformations, it's different from the DataFrames mini-language.

I think the smoothest way to handle this would be to pass all `@aes()` expressions through TidierData's `@mutate()` before passing it along to TidierPlots. This would also mean that `@aes()` would support bang-bang interpolation. I think this shouldn't be too hard to implement but would need to try it out. Of note, it wouldn't work with the plain old `aes()` function -- only for the `@aes()` macro, which supports expressions.

TidierData is a pretty light dependency since TidierPlots already depends on DataFrames, and there's not much else that TidierData adds in terms of dependency weight.

The question for @rdboyes is: what are your thoughts on TidierPlots adding TidierData as a dependency?

**Comment by rdboyes** [comment]
I think it’s no issue to add TidierData as a dependency personally. I have been thinking about how to manage the dependencies of TidierPlots better though since it is becoming pretty heavy on dependencies. Much of this is just the Makie dependency and is basically unavoidable, but setting it up to work in Pluto is currently a bit unintuitive because it “secretly depends” on WGLMakie to work correctly. I had thought about having a TidierPlotsCore that creates plot objects and different light modules to handle different types of output? Not sure how to best do it though

**Comment by kdpsingh** [comment]
I would try to avoid breaking up the package unless absolutely needed.

There is a way now in Julia to specify conditional dependencies (e.g., if Pluto is available, then install WGLMakie). I wonder if this would serve the purpose.

The simplest and key thing is just to document how to use the package with Pluto.

**Comment by rdboyes** [comment]
Progress here - in the dev build, functions can be applied inside aes (only the function version, the macro needs to be updated still): 

```
ggplot(penguins) + 
           geom_violin(aes(x = cat_inorder(:species), y = :bill_length_mm))
```
![g1](https://github.com/TidierOrg/TidierPlots.jl/assets/33524191/14d6c1a7-ec10-4094-8aaa-dee1d6108db8)


```
ggplot(penguins) + 
           geom_violin(aes(x = cat_inseq(:species), y = :bill_length_mm))
```
![g2](https://github.com/TidierOrg/TidierPlots.jl/assets/33524191/f25ed1df-b4fb-4359-8419-ce508a2079d4)

**Comment by kdpsingh** [comment]
I'll try to work on the macro version of it soon.

**Comment by rdboyes** [comment]
Operators are working in dev branch now, just have to iron out some bugs: 

```
g = ggplot(penguins) + 
  geom_point(aes(
    x = :bill_depth_mm + :bill_length_mm,
    y = :bill_length_mm, 
    color = :species)) + 
  scale_x_continuous(labels = "{:.3f}mm")
```
```
ggplot options
data: A DataFrame (333 rows, 7 columns)
xtickformat: {:.3f}mm

geom_point
data: inherits from plot
x: Calculated 
y: bill_length_mm 
x1: bill_depth_mm
x2: bill_length_mm
color: species
```

![g](https://github.com/TidierOrg/TidierPlots.jl/assets/33524191/f560145e-1d57-4c37-b00a-f6ff9fcec1f4)

**Comment by atantos** [comment]
Can't wait to hear more news on the front of debugging!

Thanks for the great work!

Alex

**Comment by rdboyes** [comment]
"Experimental support" available for this feature as of 0.6.3!: https://github.com/TidierOrg/TidierPlots.jl/commit/828db2ba211f48fd7f4a1d8e4ca2c19fc4878bad
